File: project-C++/System_I.cpp
#include "System_I.h"
using namespace std;

void CalculateCheckCode(unsigned int &Err,string& CC_,char& CheckCode)
{
	RemoveSeparator(CC_);
	if(CC_.length()!=SYS_I_UNCOMCC_LEN)
	{
		Err=1;
		return;
	}
	int num[SYS_I_UNCOMCC_LEN];//保证传入的字符串长度为9
	Base_32_legality_with_An_Arr(Err,CC_,num);//Base32正确性
	if(Err)
	{
		return;
	}

	int sum=0;
	for(int i=0;i<9;i++){				//去掉分隔符前9位
		if(i%2==0){					//从左到右奇数位上的数字（统一为32进制）
			sum+=num[i]*WEIGHT;
		}
		else{						//从左到右偶数位上的数字
			sum+=num[i];
		}
	}
	sum%=32;
	CheckCode=charNum2sym[32-sum];		//得到大写验证码
}

bool VerifyCC( string& CC ,char& CheckCode)
{
	int len=CC.length();
	char tobeComp=CC[len-1];//最后一个字符
	unsigned int Err=0;
	string CC_=CC.erase(len-1,1);
	char toComp;//正确的checkcode
	CalculateCheckCode(Err,CC_,toComp);
	if (Err)
	{
		CheckCode='#';
		return 0;
	}

	if (toComp==tobeComp)
	{
		return 1;
	}
	else {
		CheckCode=toComp;
		return 0;
	}
}

File: project-Java/SystemI.java
public class SystemI {
    final static int WEIGHT = 7;
    public static char calculateCheckCode(String CC_)throws MyException{

        if (StringManip.isDataLegal(CC_ + "0")) {//插入默认数字0,只能是0，后面计算会用到
            String CC = StringManip.RemoveSeparator(CC_+"0");//保险起见
            int sum = 0;
            for (int i = 0; i < CC.length(); i++) {
                if (i % 2 == 0) {
                    sum += Base32.symOfBase32.indexOf(CC.substring(i, i + 1)) * WEIGHT;
                } else {
                    sum += Base32.symOfBase32.indexOf(CC.substring(i, i + 1));
                }
            }
            sum %= 32;
            return Base32.symOfBase32.charAt((32 - sum)%32);
        } else {
            throw new MyException("待检验字符串不符合长度规范或格式规范-calculateCheckCode");
        }

    }

    public static boolean verifyCC(String CC)throws MyException{
        if(StringManip.isDataLegal(CC)){
            char checkcode=calculateCheckCode(CC.substring(0,CC.length()-1));
            if(checkcode-CC.charAt(CC.length()-1)==0)return true;
            else {
                System.out.println("正确字符应为："+checkcode);
                return false;
            }
        }else{
            throw new MyException("待检验字符串不符合长度规范或格式规范-verifyCC");
        }
    }
}
